.github/workflows/production-heartbeat.yml

```diff
@@ -11,28 +11,148 @@ on:
+          
+          echo "Waiting for 10 seconds..."
+          sleep 10
+          
+          echo "Listing installed extensions..."
+          "/c/Program Files/Microsoft VS Code/bin/code" --list-extensions
+        shell: bash
+
+      - name: Verify Extension on Windows
+        if: runner.os == 'Windows'
+        run: |
+          echo "Waiting for 10 seconds..."
+          sleep 10
+
+          echo "Listing installed extensions..."
+          extensions=$("/c/Program Files/Microsoft VS Code/bin/code" --list-extensions)
+
+          echo "$extensions"
+
+          if echo "$extensions" | grep -q 'salesforce.sfdx-code-analyzer-vscode'; then
+            echo "Extension 'salesforce.sfdx-code-analyzer-vscode' is installed successfully"
+          else
+            echo "Extension 'salesforce.sfdx-code-analyzer-vscode' is NOT installed"
+            exit 1
+          fi
+        shell: bash
+
+      # 3 Install VS Code and Extension on macOS
+      - name: Install VS Code on macOS
+        if: runner.os == 'macOS'
+        run: |
+          brew install --cask visual-studio-code
+
+      - name: Install Salesforce Code Analyzer Extension on macOS
+        if: runner.os == 'macOS'
+        run: |
+          code --install-extension salesforce.sfdx-code-analyzer-vscode
+
+      - name: Verify Extension Installation on macOS
+        if: runner.os == 'macOS'
+        run: |
+          if code --list-extensions | grep -q 'salesforce.sfdx-code-analyzer-vscode'; then
+            echo "Extension installed successfully"
+          else
+            echo "Extension installation failed" && exit 1
+          fi
+
+      # === Report any problems ===
+      - name: Report problems
+        # There are problems if any step failed or was skipped.
+        # Note that the `join()` call omits null values, so if any steps were skipped, they won't have a corresponding
+        # value in the string.
+        if: ${{ failure() || cancelled() }}
         shell: bash
-        run: echo "Hello World"
+        env:
+          # If we're here because steps failed or were skipped, then that's a critical problem. Otherwise it's a normal one.
+          # We can't use the `failure()` or `cancelled()` convenience methods outside of the `if` condition, hence the
+          # `contains()` calls.
+          IS_CRITICAL: ${{ contains(join(steps.*.outcome), 'failure') || contains(join(steps.*.outcome), 'skipped') }}
+          # A link to this run, so the PagerDuty assignee can quickly get here.
+          RUN_LINK: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
+
+        run: |
+
+          if [[ ${{ env.IS_CRITICAL }} == true ]]; then
+            ALERT_SEV="critical"
+            ALERT_SUMMARY="Production heartbeat script failed on ${{ runner.os }}"
+          else
+            ALERT_SEV="info"
+            ALERT_SUMMARY="Production heartbeat script succeeded with retries on ${{ runner.os }}"
```

Comment:
Does this line of code ever git hit given that the conditional on line 118 is for failure() || cancelled()?

For cancelled... do we really want to say things succeeded?

Reply:
You are right. It doesn't. But it was useful for all the tests I ran on Friday to test the pager. So, I would probably leave it. (sfdx-scanner heartbeat also follows a similar strategy. )

Reply:
I'd at least add a comment indicating that the `else` branch is impossible to hit in normal execution, maybe?